### DISCUSSION TITLE
OpenBLAS on linux: how to mitigate llama.cpp model performance degradation during chat

### DISCUSSION THREAD

**Discussion by fakezeta** [oobabooga/text-generation-webui#916] [Q&A] [answered]
Hi, 

I'm trying to setup Text Generation UI on a Free Tier OCI ARM cloud instance (4 core, 24GB RAM) with Ubuntu 22.04 image.
Everything is working fine and performance are acceptable: with in instruct mode with Vicuna template and ggml-vicuna-13b-4bit models I have more or less 2 tokens/second.

As long as the chat goes on token generation degrades a lot to 0.4 or lower: I know that it's normal due to context in the prompt. 
With Koboldcpp i mitigated this effect using the OpenBLAS library: how can I use it here?

A copy of the system info of llama.cpp model:
````
llama.cpp weights detected: models/vicuna/ggml-vicuna-13b-4bit.bin

llama_model_load: loading model from 'models/vicuna/ggml-vicuna-13b-4bit.bin' - please wait ...
llama_model_load: n_vocab = 32001
llama_model_load: n_ctx   = 2048
llama_model_load: n_embd  = 5120
llama_model_load: n_mult  = 256
llama_model_load: n_head  = 40
llama_model_load: n_layer = 40
llama_model_load: n_rot   = 128
llama_model_load: f16     = 2
llama_model_load: n_ff    = 13824
llama_model_load: n_parts = 2
llama_model_load: type    = 2
llama_model_load: ggml map size = 7759.84 MB
llama_model_load: ggml ctx size = 101.25 KB
llama_model_load: mem required  = 9807.93 MB (+ 3216.00 MB per state)
llama_model_load: loading tensors from 'models/vicuna/ggml-vicuna-13b-4bit.bin'
llama_model_load: model size =  7759.40 MB / num tensors = 363
llama_init_from_file: kv self size  = 3200.00 MB
AVX = 0 | AVX2 = 0 | AVX512 = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 0 | VSX = 0 |
````

libopenblas is in the LD path, actually in /usr/lib, installed also pylib-openblas pip package but without success.

Thank you for pointing in the right direction :)

**Comment by tye-singwa** [comment, marked as answer]
hi!
I've setup llama.cpp with OpenBLAS too, so i can help.
As i can see llama.cpp (here https://github.com/ggerganov/llama.cpp/blob/master/Makefile#L103-L105) uses flag to conditionally compile it with OpenBLAS support. 
And python bindings that used in this project compiles llama.cpp (here https://github.com/abetlen/llama-cpp-python/blob/main/CMakeLists.txt#L8) during installation.
So to enable OpenBLAS, please, try to uninstall `llama-cpp-python`, and reinstall with this flag enabled:
```shell
pip3 uninstall llama-cpp-python
LLAMA_OPENBLAS=1 pip install -r requirements.txt
```

**Reply by fakezeta** [reply]
Thank you!

**Reply by Hugo-Matias** [reply]
Is there any tradeoff that we should be aware when doing this? If it helps performance why isn't it a default? Thanks for the steps.

**Reply by electric-samurai** [reply]
Have you tried Intel MKL BLAS library?  I am hoping it is faster than OpenBLAS.

**Reply by electric-samurai** [reply]
Instructions for it are in this thread: https://github.com/ggerganov/llama.cpp/issues/719    I'm going to move this to a discussion thread, it seems signifcantly faster (if you have an intel cpu)
